### ISSUE TITLE
Where in Yii app should I be Checking scopes

### ISSUE THREAD

**Issue by mtangoo** [rhertogh/yii2-oauth2-server#12]
Hi,
thanks for the library. It saves time.
Am checking at the library and I want to be able to limit app by scopes.
In one scope I just bundle of RBAC permissions, aka roles. So I want to be able to 
check scopes to see if the requested scopes are allowed for this app.

My problem is, I cannot find where in the module am I supposed to access the scopes. Can you shed light on that?

Thanks

**Comment by rhertogh** [comment]
Hi @mtangoo,

Since you're talking about RBAC permissions I would like to emphasize first that roles and scopes are completely different thinks and should not be mixed/combined.
Scopes only come into play in delegation scenarios, and always limit what an app can do on behalf of a user: a scope cannot allow an application to do more than what the user can do. I would recommend to check out: https://auth0.com/blog/on-the-nature-of-oauth2-scopes/

To answer your question, could you please provide some more details. Are you trying to setup server-to-server authentication (with the "client_credentials" grant type) or let an end user authorize a 3rd party app (e.g. with grant type "auth_code")?

**Comment by mtangoo** [comment]
Hi @rhertogh,
Sorry for not being clear. I know the difference. But I want my scope to derive from user roles. For example, let say my app is an email app. User roles I have are: 
- user: can read/create/delete email
- admin:  can do all User can plus can create new account, delete or suspend it

Now I have two scopes in app called user_scope and manager_scope. User scope is 1:1 mapping with user, i.e can `read/create/delete email` while manager_scope allows app to only suspend account. Nothing else should this manager scope do.

Now I know I can check for permission in the controller like this
```php
if(!Yii::$app->user->can('admin')){
    throw new \Exception('You do not have priviledges to access resources')
}
```

Now am  looking on the way to check if the app is allowed to access resources based on scopes granted, like I can do to user permissions. This is where I have not found yet a solution. I was thinking of adding a class implementing  `BootstrapInterface` and attach to `beforeAction` event and check everything there. But then I'm not sure how I can access scopes for that application.

I guess my confusion is based on how do I get app credentials in the likes of the manner I can get Web User with `Yii::$app->user`.  I assume the library provides something abstracting me implementing some league interface.

The grant am using is Authorization Code Grant.

Sorry if am not yet clear, sometimes it gets hard to express the issue at hand!

**Comment by rhertogh** [comment]
So if I understand it correctly your backend application has 2 roles 'user' and 'admin'.
When a user authorizes a client (let's say a frontend application) to use the backend application all calls that are made by that client will be automatically done with that user's identity.

Let's say that user Alice is "admin" and user Bob is a regular "user". The backend application we abbreviate as BEA and the frontend application as FEA.

When Alice authorizes the FEA via the Authorization Code Grand the FEA can use the bearer token to make API calls against the BEA. For all these calls the `Yii::$app->user->identity` will be set to the user identity of Alice.
When Bob would authorize the FEA all calls with his bearer token by the FEA will set the `Yii::$app->user->identity` to Bob's identity.

Since the identity is set, you can simply use `Yii::$app->user->can('role_you_want_to_check')` anywhere you want.
E.g. `Yii::$app->user->can('admin')` would return `true` for Alice and `false` for Bob.

Note that until now "scope" is never used and probably is not needed.
Only when we have multiple clients, let's say FEA1 and FEA2, that need a different access scope for the *same* user, scopes come into play.

Let's say that Bob (with the "user" role) wants to authorize two different clients:
 - FEA1 to read, create and delete emails on his behalf.
 - FEA2 to only create emails on his behalf.

In that case you would need to define the scopes 'read_email', 'create_email' and 'delete_email' and assign them to the different clients.
Then Bob can specify which scopes the client may use during the authorization of the client.
Note that another user, let's say Charles (who also has the role "user"), can allow different scopes for the same clients, e.g.:
 - FEA1 may read and create emails on his behalf.
 - FEA2 may create and delete emails  on his behalf.

When scopes are used you can use `Yii::$app->getModule('oauth2')->requestHasScope('scope_you_want_to_check')` to see if the client that makes the request has the required scope.

So in the case of Bob and Charles:
 - `Yii::$app->user->can('admin')` would be `false` for both.
 - `Yii::$app->user->can('user')` would be `true` for both.
 - And when checking the scopes (B = Bob, C = Charles):
   |       |     'read_email'     |   'create_email'   |   'delete_email'    |
   |-------|:--------------------:|:------------------:|:-------------------:|
   | FEA1  |  B:`true`, C:`true`  | B:`true`, C:`true` | B:`true`, C:`false` |
   | FEA2  | B:`false`, C:`false` | B:`true`, C:`true` | B:`false`, C:`true` |

To summarize:
- Roles define what the user may do.
- Scopes limit what a client may do on behalf of the user.
- If you don't need to limit what a client may do on behalf of a user, you probably don't need scopes.
- If you do need scopes, you most likely need to check both the user Role and Scope. E.g.
   ```php
   class EmailController extends Controller
   {
       public function actionCreate()
       {
           if (
               !Yii::$app->user->can('user')
               || !Yii::$app->getModule('oauth2')->requestHasScope('create_email')
           ) {
               throw new UnauthorizedHttpException();
           }
           // ...
   ```

**Comment by mtangoo** [comment]
Wow!
That was detailed. I think would be good to have it in docs somewhere.
I just noticed I was looking for `Yii::$app->getModule('oauth2')->requestHasScope('scope_you_want_to_check')`
The missing piece... THANK YOU a million!

**Comment by rhertogh** [comment]
> I think would be good to have it in docs somewhere.

#17
